### ISSUE TITLE
NaN when training on TPUs with large batch size

### ISSUE THREAD

**Issue by bram-w** [ronghanghu/moco_v3_tpu#1]
Hello, thank you so much for this repository; my team is exploring adopting TPUs and this as a base has been very valuable.

I've run into an issue, however, with scaling up to large batch sizes. Once the batch size increases into the thousands the cross entropy loss gives a NaN, example output below from the call

`python3 -m torch_xla.distributed.xla_dist --tpu=$TPU_NAME    --restart-tpuvm-pod-server  --env XLA_USE_BF16=0   -- python3 /my_simclr_vit_profiling/run_simclr_vit.py   device=xla   ckpt_dir=$SAVE_DIR fake_data=True lr=0 weight_decay=0
`

```
2022-04-27 20:20:45 172.16.97.60 [0] training begins (note that the first few XLA iterations are very slow due to compilation)
2022-04-27 20:20:45 172.16.97.60 [0] starting epoch 1
2022-04-27 20:26:58 172.16.97.60 [0] epoch 1 step 50, lr: 0.0000, loss: nan, loss (avg): nan, loss (median): nan
```

As opposed to
`python3 -m torch_xla.distributed.xla_dist --tpu=$TPU_NAME    --restart-tpuvm-pod-server  --env XLA_USE_BF16=0   -- python3 /my_simclr_vit_profiling/run_simclr_vit.py   device=xla   ckpt_dir=$SAVE_DIR fake_data=True batch_size=512 lr=0 weight_decay=0`

yielding

```
2022-04-27 20:20:33 172.16.96.39 [0] training begins (note that the first few XLA iterations are very slow due to compilation)
2022-04-27 20:20:33 172.16.96.39 [0] starting epoch 1
2022-04-27 20:25:20 172.16.96.39 [0] epoch 1 step 50, lr: 0.0000, loss: 6.9305, loss (avg): 6.9305, loss (median): 6.9305
2022-04-27 20:25:32 172.16.96.39 [0] epoch 1 step 100, lr: 0.0000, loss: 6.9305, loss (avg): 6.9305, loss (median): 6.9305
2022-04-27 20:25:43 172.16.96.39 [0] epoch 1 step 150, lr: 0.0000, loss: 6.9305, loss (avg): 6.9305, loss (median): 6.9305
```

The loss is NaN from the first step. I noticed it when modifying the code to run on a CLIP objective. The logits at https://github.com/ronghanghu/simclr_vit_profiling/blob/simclr_vit_release/losses.py#L47 have reasonable values (no NaNs), but the output from the cross entropy loss is NaN.

Do you have any advice or experience with this from your work? Thank you so much in advance!

Background info:
Running on a v3-32 (NaN emerges with v3-256 as well)
tpu runtime version tpu-vm-pt-1.11

**Comment by ronghanghu** [comment]
Hi, I saw that you are running with fake data. Does NaN still occur frequently if you switch to real data?

Fake data is expected to cause NaNs because all the samples in a batch are the same fake sample, making batch normalization impossible to work (because the variance across batch is 0)

**Comment by bram-w** [comment]
It does, but I think I've diagnosed it as a problem with the TPU software: 1.10 and v2-alpha didn't get a NaN while 1.11 did.

Just to double-check, which software were you on? And thanks for the fast response

**Comment by ronghanghu** [comment]
I see. I've been using `tpu-vm-pt-1.10` (instead of `v2-alpha`) as my TPU VM runtime, which contains the PyTorch and PyTorch XLA 1.10 builds.

**Comment by bram-w** [comment]
Excellent. Thank you for the information (even though it looks like it was only tangentially related to your code). Closing issue now, much appreciated
